### PR TITLE
Add ERC20 opt-in migration contract

### PR DESCRIPTION
**Pull request by facuspagnuolo**
Fixes https://github.com/zeppelinos/zos-lib/issues/185

# 🚀 Description

This PR adds a new contract to migrate ERC20 tokens with an opt-in strategy. 
Additionally, I'm extracting the behavior expected of the `StandardToken` so we can start reusing it in different test files.

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

File: test/token/ERC20/BurnableToken.test.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import shouldBehaveLikeBurnableToken from './BurnableToken.behaviour';
+import shouldBehaveLikeBurnableToken from './behaviors/BurnableToken.behaviour';
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply test/token/ERC20/BurnableToken.test.js]: I'm removing these changes to avoid conflicts

File: contracts/token/ERC20/MigratableERC20.sol [hunk 2]
```diff
@@ -56,17 +55,12 @@ contract OptInERC20Migration is StandardToken {
    */
   function migrateTokenTo(address _to, uint256 _amount) public {
     _mint(_to, _amount);
-    legacyToken.transferFrom(msg.sender, BURN_ADDRESS, _amount);
+    legacyToken.safeTransferFrom(msg.sender, BURN_ADDRESS, _amount);
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply contracts/token/ERC20/MigratableERC20.sol]: I see your point but note that we are not decreasing and incrementing the same balance here. Given that we need to call a foreign contract, I went with the (check-effects-interactions)[https://solidity.readthedocs.io/en/v0.4.24/security-considerations.html#use-the-checks-effects-interactions-pattern] pattern instead.

**Comment by facuspagnuolo** [comment]
Comments addressed, please take another look @spalladino

File: contracts/token/ERC20/MigratableERC20.sol [hunk 3]
```diff
@@ -0,0 +1,67 @@
+pragma solidity ^0.4.24;
+
+import "./ERC20.sol";
+import "./SafeERC20.sol";
+
+
+/**
+ * @title MigratableERC20
+ * @dev This strategy carries out an optional migration of the token balances. This migration is performed and paid for
+ * @dev by the token holders. The new token contract starts with no initial supply and no balances. The only way to
```
Comment (frangio) [inline contracts/token/ERC20/MigratableERC20.sol]: Please remove the `@dev` tag from all but the first line here.

File: contracts/token/ERC20/MigratableERC20.sol [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * @dev "mint" the new tokens is for users to "turn in" their old ones. This is done by first approving the amount they
+ * @dev want to migrate via `ERC20.approve(newTokenAddress, amountToMigrate)` and then calling a function of the new
+ * @dev token called `migrateTokens`. The old tokens are sent to a burn address, and the holder receives an equal amount
+ * @dev in the new contract.
```
Comment (frangio) [inline contracts/token/ERC20/MigratableERC20.sol]: These inline docs explain how the contract works but not how it's used. I'd add a note saying at least that it's an abstract contract that must be combined with `StandardToken` or the user's own token implementation (with a `_mint` internal function), and if we have a link to a guide for how to use it I'd put that here as well!

**Comment by facuspagnuolo** [comment]
@frangio I had already changed all the `import` statements in favour of `require` ones. I'm not finding any `import`s left, please point them out in case you see one.

File: contracts/token/ERC20/MigratableERC20.sol [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    migrateTokenTo(msg.sender, _amount);
+  }
+
+  /**
+   * @dev Burns a given amount of the old token contract for a token holder and mints the same amount of
+   * @dev new tokens for a given recipient address
+   * @param _amount uint256 representing the amount of tokens to be migrated
+   * @param _to address the recipient that will receive the new minted tokens
+   */
+  function migrateTokenTo(address _to, uint256 _amount) public {
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply contracts/token/ERC20/MigratableERC20.sol]: I guess I'm not following, do you prefer holding all the logic together in `migrate` or separately in some way?

File: contracts/drafts/ERC20Migrator.sol [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * To opt-in, users must approve for this contract the number of tokens they
+ * want to migrate. Once the allowance is set up, anyone can trigger the
+ * migration to the new token contract. In this way, token holders "turn in"
+ * their old balance and will be minted an equal amount in the new token.
+ * The new token contract must be mintable. For the precise interface refer to
+ * OpenZeppelin's ERC20Mintable, but the only functions that are needed are
+ * `isMinter(address)` and `mint(address, amount)`. The migrator will check
+ * that it is a minter for the token.
+ * The balance from the legacy token will be transfered to the migrator, as it
+ * is migrated, and remain here forever.
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: Consider replacing _here_ for _there_

File: contracts/drafts/ERC20Migrator.sol [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * Although this contract can be used in many different scenarios, the main
+ * motivation was to provide a way of how an ERC20 token can be migrated to an
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: Consider replacing _a way of how an ERC20 token can be migrated to an_ for _a way to migrate ERC20 tokens into an_

File: contracts/drafts/ERC20Migrator.sol [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  function beginMigration(ERC20Mintable newToken) public {
+    require(_newToken == address(0));
+    require(newToken != address(0));
+    require(newToken.isMinter(this));
+
+    _newToken = newToken;
+  }
+
+  /**
+   * @dev Burns a given amount of the old token contract for a token holder and mints the same amount of
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: Consider rephrasing; tokens aren't burned, but instead transfered here.

File: contracts/drafts/ERC20Migrator.sol [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+   * @dev new tokens for a given recipient address
+   * @param account uint256 representing the amount of tokens to be migrated
+   * @param amount address the recipient that will receive the new minted tokens
+   */
+  function migrate(address account, uint256 amount) public {
+    _newToken.mint(account, amount);
+    _legacyToken.safeTransferFrom(account, this, amount);
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: I'd swap these two statements, for clarity (i.e. first transfer to migrator, then mint new tokens).

File: contracts/drafts/ERC20Migrator.sol [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  /**
+   * @dev Migrates the total balance of the token holder to this token contract
+   * @dev This function will burn the old token balance and mint the same balance in the new token contract
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: Again, _burn_ is not accurate.

File: contracts/drafts/ERC20Migrator.sol [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+   * @param account uint256 representing the amount of tokens to be migrated
+   * @param amount address the recipient that will receive the new minted tokens
+   */
+  function migrate(address account, uint256 amount) public {
+    _newToken.mint(account, amount);
+    _legacyToken.safeTransferFrom(account, this, amount);
+  }
+
+  /**
+   * @dev Migrates the total balance of the token holder to this token contract
```
Comment (nventuro) [inline contracts/drafts/ERC20Migrator.sol]: It doesn't transfer the total balance, but the total _approved_ balance

File: test/drafts/ERC20Migrator.test.js [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  .should();
+
+contract('ERC20Migrator', function ([_, owner, recipient, anotherAccount]) {
+  const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+  const totalSupply = 200;
+
+  beforeEach('deploying tokens', async function () {
+    this.legacyToken = await ERC20Mock.new(owner, totalSupply);
+    this.migrator = await ERC20Migrator.new(this.legacyToken.address);
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: Consider adding a test with a null legacy token.

File: test/drafts/ERC20Migrator.test.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    this.newToken = await ERC20Mintable.new();
+  });
+
+  describe('beginMigration', function () {
+    it('rejects zero address for token', async function () {
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: As per #1157, this should read something like 'reverts with a null new token address'

File: test/drafts/ERC20Migrator.test.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+    it('reverts the second time it is called', async function () {
+      await this.newToken.addMinter(this.migrator.address);
+      await this.migrator.beginMigration(this.newToken.address);
+      await assertRevert(this.migrator.beginMigration(this.newToken.address));
+    });
+  });
+
+  describe('once migration began', function () {
+    beforeEach('', async function () {
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: Remove the empty block description

File: test/drafts/ERC20Migrator.test.js [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+      const baseAmount = totalSupply;
+
+      describe('when the approved balance is higher or equal to the owned balance', function () {
+        const amount = baseAmount;
+
+        beforeEach('approving the whole balance to the new contract', async function () {
+          await this.legacyToken.approve(this.migrator.address, amount, { from: owner });
+        });
+
+        beforeEach('migrating token', async function () {
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: This `describe` has two `beforeEach` blocks, which are not independent: I'm not sure if Mocha specifies in which order they'll be run (or even if both will be run), but we should avoid this situation IMO.

File: test/drafts/ERC20Migrator.test.js [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+  describe('once migration began', function () {
+    beforeEach('', async function () {
+      await this.newToken.addMinter(this.migrator.address);
+      await this.migrator.beginMigration(this.newToken.address);
+    });
+
+    describe('migrateAll', function () {
+      const baseAmount = totalSupply;
+
+      describe('when the approved balance is higher or equal to the owned balance', function () {
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: nit: this only tests for equal to balance, not higher

File: test/drafts/ERC20Migrator.test.js [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    });
+
+    describe('migrate', function () {
+      const baseAmount = 50;
+
+      beforeEach('approving tokens to the new contract', async function () {
+        await this.legacyToken.approve(this.migrator.address, baseAmount, { from: owner });
+      });
+
+      describe('when the amount is lower or equal to the one approved', function () {
```
Comment (nventuro) [inline test/drafts/ERC20Migrator.test.js]: This only tests for equal

File: contracts/drafts/ERC20Migrator.sol [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * @dev Burns a given amount of the old token contract for a token holder and mints the same amount of
+   * @dev new tokens for a given recipient address
+   * @param account uint256 representing the amount of tokens to be migrated
+   * @param amount address the recipient that will receive the new minted tokens
+   */
+  function migrate(address account, uint256 amount) public {
+    _newToken.mint(account, amount);
+    _legacyToken.safeTransferFrom(account, this, amount);
```
Comment (spalladino) [inline contracts/drafts/ERC20Migrator.sol]: Since minting happens before transferring, there could be some odd behaviours if the legacy and new token are the same (you could migrate a near-infinite amount). But I'm not sure that's a scenario we even want to consider.

File: contracts/drafts/ERC20Migrator.sol [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    _newToken = newToken;
+  }
+
+  /**
+   * @dev Burns a given amount of the old token contract for a token holder and mints the same amount of
+   * @dev new tokens for a given recipient address
+   * @param account uint256 representing the amount of tokens to be migrated
+   * @param amount address the recipient that will receive the new minted tokens
+   */
+  function migrate(address account, uint256 amount) public {
```
[the remark this answers is not in the document]
Comment (nventuro) [reply contracts/drafts/ERC20Migrator.sol]: The (soft) policy we currently have is that no-op transactions (barring events being emitted) are allowed, to prevent burdening the caller with the responsibility of checking for non-zero.

File: contracts/drafts/ERC20Migrator.sol (hunk 18 above)
Comment (nventuro) [reply contracts/drafts/ERC20Migrator.sol]: Hm, I guess we could make sure `legacyToken` and `newToken` are not the same contract?

Comment (spalladino) [reply contracts/drafts/ERC20Migrator.sol]: Not sure if there is any point in it, as one could be a proxy, so you have no way to compare. I mentioned it just because it's generally a good practice to decrement (ie transfer) before incrementing (ie minting).

Comment (nventuro) [reply contracts/drafts/ERC20Migrator.sol]: I thought about that, but thought there was value in preventing a simple user error.

**Comment by frangio** [comment]
Tests passed locally.
